### ISSUE TITLE
Composer compatible builds 

### ISSUE THREAD

**Issue by jameskraus** [versionpress/versionpress#939]
Right now, when builds are produced, the plugin is stuffed in a "versionpress" directory and the composer.json file is deleted. The builds can be made composer compatible by 

- Placing plugin files in root of .zip
- Keeping the composer.json file

So, how about it?

**Comment by jameskraus** [comment]
And, if possible, can `"type": "wordpress-plugin"` be added to the composer.json file? It would help composer understand where to place the plugin in bedrock setups.

**Comment by LukeAbell** [comment]
+1

**Comment by borekb** [comment]
Jan was experimenting with this, see https://github.com/wp-cpm/versionpress. If we were to integrate this into the main project, how would it work? Would we build another ZIP file on every release, attach it to the Releases section and instruct Packagist to get it from there?

Do we want to do it in the first place? To explain this, we're hesitant to make the installation too easy for now as VersionPress is still not compatible with many plugins and can break the site in the worst case. I understand that Composer users might be more tech-savvy though and know what they are doing..

**Comment by jameskraus** [comment]
I don't think it's possible to instruct packagist to grab it from the releases. From what I know of packagist, it's only possible to point it at the entire repository, and since versionpress has build steps required before it can be used, that would probably not work.

The easiest thing for now would be to keep files in the root of the build zip, retain composer.json (with appropriate type) and release builds which people can point at manually from their own composer.json files. 

For example, I'm working on a bedrock site which includes this in the composer.json:

`"repositories": [
         ...
	{
		"type": "package",
		"package": {
			"name": "jameskraus/versionpress",
			"type": "wordpress-plugin",
			"version": "3.0.0",
			"dist": {
				"url": "https://github.com/jameskraus/versionpress/releases/download/3.0-beta/versionpress.zip",
				"type": "zip"
			}
		}
	}
  ],
  "require": {
        ....
	"jameskraus/versionpress": "3.0.0"
  },
  ....

`

This allows me to include my most recent build and when my bedrock site builds, my build of versionpress is automatically installed.

**Comment by borekb** [comment]
This repo would become quite a mess if we moved everything to the root so it'll probably stay the way it is, however, we could have another repo which would be maintained to automatically mirror the main one, with the structure you suggested. Scheduling this for 4.0, feel free to create such mirroring repo if you need it sooner :)

BTW https://travis-ci.org/jameskraus/versionpress/ is cool! We need to do a proper CI one day and had already experimented with Docker for full-blown Selenium tests but didn't manage to get it finished yet. Travis would be nice if it supports Selenium, plus the ability to create temporary WP testing sites (which I doubt but maybe Travis got more powerful since the last time I looked).

**Comment by jameskraus** [comment]
> This repo would become quite a mess if we moved everything to the root so it'll probably stay the way it is, however, we could have another repo which would be maintained to automatically mirror the main one, with the structure you suggested.

Oh yeah, to be clear, I'm not suggesting moving everything to the root of this repo. I suggesting moving all the files in the build .zip from `versionpresss/*` to the root. 

i.e. from this:

```
versionpress-3.0-beta.zip
└───versionpress
    │   .htaccess
    │   bootstrap.php
    │   uninstall.php
    │   versionpress-functions.php
    │   versionpress-nginx.conf
    │   versionpress.php
    │   vpconfig.defaults.yml
    │   vpconfig.yml
    │   web.config
    │
    ├───admin
    ├───log
    ├───src
    ├───temp
    └───vendor
```
to this:
```
versionpress-3.0-beta.zip
│   .htaccess
│   bootstrap.php
│   composer.json
│   uninstall.php
│   versionpress-functions.php
│   versionpress-nginx.conf
│   versionpress.php
│   vpconfig.defaults.yml
│   vpconfig.yml
│   web.config
│
├───admin
├───log
├───src
├───temp
└───vendor
```


> Travis would be nice if it supports Selenium, plus the ability to create temporary WP testing sites (which I doubt but maybe Travis got more powerful since the last time I looked).

I'm pretty sure you can either run selenium directly on travis or you can run it though sauce labs. Travis seems to [recommend using sauce labs](https://docs.travis-ci.com/user/sauce-connect/) for it and they are [free for open source projects](https://saucelabs.com/opensauce/).

**Comment by borekb** [comment]
@jameskraus I'd like to better understand what the goal is here. The main ZIP we distribute uses the same structure as plugins on wordpress.org which leads to a natural installation path of `wp-content/plugins/versionpress`. If we move everything to the root of the ZIP file, the installation path would change to something like `wp-content/plugins/versionpress-3.0-beta`. This might not be causing any technical trouble but looks slightly weird.

So we could add another ZIP that has the structure you propose but what's the use case for it? Are you going to install this ZIP through Composer somehow? If so, how exactly? Does it benefit you that there is another ZIP when you can easily get the contents of it from the current one already?

Thanks for clearing that up.

**Comment by jameskraus** [comment]
> If we move everything to the root of the ZIP file, the installation path would change to something like wp-content/plugins/versionpress-3.0-beta. This might not be causing any technical trouble but looks slightly weird.

Huh, I didn't know that was how WordPress handled plugin .zip's. Then I guess it makes the most sense to keep the build zip the way it is. 

> So we could add another ZIP that has the structure you propose but what's the use case for it? Are you going to install this ZIP through Composer somehow? If so, how exactly? Does it benefit you that there is another ZIP when you can easily get the contents of it from the current one already?

I'm trying to install VP on a composer-based site, which requires either a .zip with the composer.json in the root (decent stopgap), or for the package to be listed on wpackagist or packagist (much easier to use). 

I've got to be real though, I found there were some major incompatibilities between my bedrock site and versionpress, even after I got the install working, so I'm probably not going to push this issue in the future.

**Comment by borekb** [comment]
We plan to look into Composer workflows in the near future so your feedback will certainly be helpful – thank you! I'm keeping this open and scheduled for 4.0.

**Comment by landsman** [comment]
It will be good!

**Comment by seagyn** [comment]
Could you not have a separate build that pushes the plugin directory to a separate repo? Only running on tagged releases etc.

**Comment by borekb** [comment]
Yes, that's most likely the way to go. The core team will need to focus on feature development first, would there be any volunteers to help set up the separate repo and the necessary build scripts?

**Comment by seagyn** [comment]
I would be interested in helping to do that. What would something like that look like?

**Comment by borekb** [comment]
I think a good start would be a manual script that would publish the latest tag from `versionpress/versionpress` to another repo (can be your own to begin with; we'd then probably provide some official location under the `versionpress` GitHub organization). It would need to do something similar to the [production build step](https://github.com/versionpress/versionpress/blob/8244d42a267d2041c09eaf08031aa5cdaeb26a4b/docs/Dev-Setup.md#production-build).
